From 4f4601ce18d1571cb2355c20376e559113749128 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Tue, 10 Sep 2024 17:07:46 +0800
Subject: [PATCH 121/236] mtk: hostapd: remove sta_info from partner links when
 deauthenticating inactive STA

Originally only the sta_info of the associated hapd will be remove,
which might cause problem when STA triggers re-authentication.

This commit removes sta_info from parnter links when deauthenticating
inactive STA so that the STA is removed on a per-MLD basis.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/sta_info.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index 62846d81c..d19d72d5f 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -49,6 +49,7 @@ static void ap_sta_disassoc_cb_timeout(void *eloop_ctx, void *timeout_ctx);
 static void ap_sa_query_timer(void *eloop_ctx, void *timeout_ctx);
 static int ap_sta_remove(struct hostapd_data *hapd, struct sta_info *sta);
 static void ap_sta_delayed_1x_auth_fail_cb(void *eloop_ctx, void *timeout_ctx);
+static void ap_sta_remove_link_sta(struct hostapd_data *hapd, struct sta_info *sta);
 
 int ap_for_each_sta(struct hostapd_data *hapd,
 		    int (*cb)(struct hostapd_data *hapd, struct sta_info *sta,
@@ -572,6 +573,8 @@ void ap_handle_timer(void *eloop_ctx, void *timeout_ctx)
 			       HOSTAPD_LEVEL_INFO, "deauthenticated due to "
 			       "local deauth request");
 		hostapd_ubus_notify(hapd, "local-deauth", sta->addr);
+		if (ap_sta_is_mld(hapd, sta))
+			ap_sta_remove_link_sta(hapd, sta);
 		ap_free_sta(hapd, sta);
 		return;
 	}
@@ -730,6 +733,8 @@ skip_poll:
 			hapd, sta,
 			WLAN_REASON_PREV_AUTH_NOT_VALID);
 		hostapd_ubus_notify(hapd, "inactive-deauth", sta->addr);
+		if (ap_sta_is_mld(hapd, sta))
+			ap_sta_remove_link_sta(hapd, sta);
 		ap_free_sta(hapd, sta);
 		break;
 	}
-- 
2.45.2

